File: providers/codex.py
"""Codex CLI provider."""

from core import config
from providers.models import ProviderMeta, ProviderResult
from providers.parsers.codex import extract_session_id

from .base import CLIProvider

_extract_session_id = extract_session_id


def permission_args(mode: str) -> list[str]:
    """Безопасные режимы для неинтерактивного codex exec."""
    if mode == "read-only":
        return ["--sandbox", "read-only", "-c", "approval_policy=never"]
    if mode == "workspace":
        return ["--sandbox", "workspace-write", "-c", "approval_policy=never"]
    return []


class CodexProvider(CLIProvider):
    provider_name = "codex"

    def env(self) -> dict:
        e = super().env()
        e["CODEX_HOME"] = str(self.cli_home)
        return e

    async def run(self, prompt, cwd, session_id, model, attachments=None, progress=None):
        # Codex пока не стримит — progress игнорируется
        full_prompt = prompt or ""
        if attachments:
            paths = "\n".join(f"- {p}" for p in attachments)
            full_prompt += f"\n\n[Прикреплённые файлы]\n{paths}\n"

        if session_id:
            argv = ["codex", "exec", "resume", session_id, "--skip-git-repo-check"]
        else:
            argv = ["codex", "exec", "--skip-git-repo-check"]
        if model:
            argv += ["-c", f"model={model}"]
        argv += permission_args(self.permission_mode)
        # инструкция языка через config override
        codex_instructions = (
            config.RU_SYSTEM_INSTRUCTION
            + "\n\nShell-команды начинай с rtk для сжатия вывода: "
            "rtk git status, rtk ls, rtk grep ..., rtk pytest ... "
            "Для составных команд: rtk sh -c '...'. "
            "Это экономит контекстные токены."
        )
        argv += ["-c", f"instructions={codex_instructions!r}"]
        argv += [full_prompt]

        rc, out, err = await self._exec(argv, cwd)
        if rc != 0:
            raise RuntimeError(f"codex failed (rc={rc}): {err[:2000]}")

        result = ProviderResult(
            text=out.strip(),
            session_id=_extract_session_id(out, err, session_id),
            meta=ProviderMeta(),
        )
        return result.as_tuple()

File: cli/src/providers/codex.ts
import { spawn } from 'node:child_process';
import path from 'node:path';
import fs from 'node:fs';
import type { Account, ProgressCallback, Provider, ProviderResult } from '../types.js';

/** Извлекает session ID из вывода codex (эвристика). */
function extractSessionId(out: string, err: string, fallback: string | null): string | null {
  const combined = out + '\n' + err;
  const match = combined.match(/session[_\s-]*id[:\s]+([a-f0-9-]{8,})/i);
  return match ? match[1] : fallback;
}

export class CodexProvider implements Provider {
  constructor(private account: Account) {}

  async run(
    prompt: string,
    cwd: string,
    sessionId: string | null,
    model: string | null,
    progress: ProgressCallback,
    attachments?: string[],
  ): Promise<ProviderResult> {
    const cliHome = this.account.cli_home_path;
    fs.mkdirSync(cliHome, { recursive: true });

    const env: Record<string, string> = {
      ...process.env as Record<string, string>,
      CODEX_HOME: cliHome,
    };

    const instructions =
      'Отвечай на русском. Будь краток. Shell-команды начинай с rtk для сжатия вывода.';

    let args: string[];
    if (sessionId) {
      args = ['codex', 'exec', 'resume', sessionId, '--skip-git-repo-check'];
    } else {
      args = ['codex', 'exec', '--skip-git-repo-check'];
    }
    if (model) args.push('-c', `model=${model}`);
    args.push('-c', `instructions=${JSON.stringify(instructions)}`);
    if (attachments) {
      for (const img of attachments) args.push('-i', img);
    }
    args.push(prompt);

    const child = spawn(args[0], args.slice(1), {
      cwd,
      env,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    (globalThis as any).__ha_process = child;

    let stdout = '';
    let stderr = '';
    child.stdout?.on('data', (chunk: Buffer) => {
      const text = chunk.toString();
      stdout += text;
      progress({ type: 'text', text });
    });
    child.stderr?.on('data', (chunk: Buffer) => { stderr += chunk.toString(); });

    const code = await new Promise<number>((resolve) => {
      child.on('close', (c) => resolve(c ?? 1));
    });

    if (code !== 0 && !stdout.trim()) {
      throw new Error(`codex failed (rc=${code}): ${stderr.slice(0, 500)}`);
    }

    return {
      text: stdout.trim(),
      sessionId: extractSessionId(stdout, stderr, sessionId),
    };
  }
}